z3c/dependencychecker/db.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class ImportsDatabase(object):
+
+    def __init__(self):
+        self._requirements = []
+        self._extras_requirements = {}
+
+    def set_requirements(self, requirements_list):
+        for requirement in requirements_list:
+            if self._requirement_already_exists(requirement):
+                logger.warn(
+                    '"%s" is declared twice on setup.py',
+                    requirement.name,
+                )
+            else:
+                self._requirements.append(requirement)
+
+    def _requirement_already_exists(self, requirement):
+        names = self.get_requirements_names()
+        return bool(requirement.name in names)
+
+    def get_requirements_names(self):
+        return [
+            x.name
+            for x in self._requirements
+        ]
+
+    def set_extra_requirements(self, extra_name, dotted_names):
+        """A bit of extra work needs to be done as pkg_resources API returns
+        all common requirements as well as the extras when asking for an extra
+        requirements.
+        """
+        only_extra_dotted_names = self._remove_duplicates(dotted_names)
+        if extra_name in self.get_extras():
+            new_dotted_names = self._only_new_requirements(
+                extra_name,
+                dotted_names,
+            )
+            self._extras_requirements[extra_name] += new_dotted_names
+
+            logger.warn(
+                'extra requirement "%s" is declared twice on setup.py',
+                extra_name,
+            )
+        else:
+            self._extras_requirements[extra_name] = only_extra_dotted_names
+
+    def _remove_duplicates(self, imports, other=None):
+        other_names = other
+        if other is None:
+            other_names = self.get_requirements_names()
+
+        filtered_list = [
+            x
+            for x in imports
+            if x.name not in other_names
+        ]
+
+        return filtered_list
+
+    def _only_new_requirements(self, extra_name, dotted_names):
+        current_names = self.get_requirements_names_for_extra(extra_name)
+        only_new = self._remove_duplicates(
```

Comment:
Could you do this with a python set? Duplication handling, subtracting sets, it is all there.

Reply:
As I mentioned above on other comments I (ab)use(?) iterators, so they need to be consumed.

Now I'm wondering if for these specific cases, where the total number of elements (here talking about setup.py requirements and extra requirements) should I rather just return the whole list rather than passing iterators around.

I looks fancier, but to not much gain (at least in this specific code), on modules.py I would keep the iterators, as there can be potentially super big packages with hundreds of files and files with hundreds of imports on them.

I'm keeping them by now, but while fixing all your feedback, if I grew tired of it I might end up changing it.